### DISCUSSION TITLE
Error:  Discovered incompatible solidity versions in following

### DISCUSSION THREAD

**Discussion by Killpit** [Cyfrin/security-and-auditing-full-course-s23#78] [Q&A]
Hi there,

Whenever I try to use contracts, I get this error and even though I tried to look for similar searches, I failed to find a solution to end this problem for good and now all forge commands are not working. Here's the GitHub to check the code and before that I updated contracts and used [profile.solc] = 0.8.20

```
macbook-pro:prereqs atatekeli$ forge build
[⠃] Compiling...2024-01-07T16:36:08.170101Z ERROR ethers_solc::resolver: failed to resolve versions
[⠊] Compiling...
Error: 
Discovered incompatible solidity versions in following
: script/Counter.s.sol (^0.8.20) imports:
    lib/forge-std/src/Script.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/console.sol (>=0.4.22 <0.9.0)
    lib/forge-std/src/console2.sol (>=0.4.22 <0.9.0)
    lib/forge-std/src/safeconsole.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdChains.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdCheats.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdJson.sol (>=0.6.0 <0.9.0)
    lib/forge-std/src/StdMath.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdStorage.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdStyle.sol (>=0.4.22 <0.9.0)
    lib/forge-std/src/StdUtils.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Base.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdStorage.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/console2.sol (>=0.4.22 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/interfaces/IMulticall3.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/mocks/MockERC20.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/mocks/MockERC721.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdStorage.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
Discovered incompatible solidity versions in following
: src/Counter.sol (^0.8.20) imports:
Discovered incompatible solidity versions in following
: src/MyToken.sol (^0.8.20) imports:
    lib/openzeppelin-contracts/contracts/token/ERC20/ERC20.sol (^0.8.20)
    lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol (^0.8.20)
    lib/openzeppelin-contracts/contracts/token/ERC20/extensions/IERC20Metadata.sol (^0.8.20)
    lib/openzeppelin-contracts/contracts/utils/Context.sol (^0.8.20)
    lib/openzeppelin-contracts/contracts/interfaces/draft-IERC6093.sol (^0.8.20)
    lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol (^0.8.20)
Discovered incompatible solidity versions in following
: test/Counter.t.sol (^0.8.20) imports:
    lib/forge-std/src/Test.sol (>=0.6.2 <0.9.0)
    src/Counter.sol (^0.8.20)
    lib/forge-std/src/console.sol (>=0.4.22 <0.9.0)
    lib/forge-std/src/console2.sol (>=0.4.22 <0.9.0)
    lib/forge-std/src/safeconsole.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdAssertions.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdChains.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdCheats.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdError.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdInvariant.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdJson.sol (>=0.6.0 <0.9.0)
    lib/forge-std/src/StdMath.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdStorage.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdStyle.sol (>=0.4.22 <0.9.0)
    lib/forge-std/src/StdUtils.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Base.sol (>=0.6.2 <0.9.0)
    lib/forge-std/lib/ds-test/src/test.sol (>=0.5.0)
    lib/forge-std/lib/ds-test/src/test.sol (>=0.5.0)
    lib/forge-std/src/StdMath.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdStorage.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/console2.sol (>=0.4.22 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/interfaces/IMulticall3.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/mocks/MockERC20.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/mocks/MockERC721.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdStorage.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
```

Here's the GitHub link to my repo

https://github.com/Killpit/web3-security-course/tree/main/prereqs

**Comment by DevSwayam** [comment]
Hey, it's about incompatible versions. Can you please run `foundryup` again in the repo or try to clone the repo again? I think the issue is regarding a contract whose version isn't matching with all of your files.

**Comment by Killpit** [comment]
Whenever I run `foundryup` it says `command not found`

**Comment by Killpit** [comment]
My plan is to delete everything about Foundry and re-install it

**Reply by Horpheyhemmy-18** [reply]
Try checking if foundry is till installed.
Run
``` forge --version```

You should have an output indicating the forge version .
If not run ``` foundryup``` to update the forge version

**Reply by Killpit** [reply]
forge 0.2.0 (98fa895 2022-09-27T22:27:26.550668Z)

**Comment by Killpit** [comment]
Here are the paths sorted alphabetically, presented one by one:

/Library/Apple/usr/bin
/Library/Frameworks/Python.framework/Versions/2.7/bin
/Library/Frameworks/Python.framework/Versions/3.7/bin
/Library/Frameworks/Python.framework/Versions/3.7/bin
/Library/Frameworks/Python.framework/Versions/3.7/bin
/Library/Frameworks/Python.framework/Versions/3.9/bin
/Library/Frameworks/Python.framework/Versions/3.9/bin
/Library/Frameworks/Python.framework/Versions/3.9/bin
/Library/Frameworks/Python.framework/Versions/3.9/bin
/Library/Frameworks/Python.framework/Versions/3.9/bin
/Library/Frameworks/Python.framework/Versions/3.10/bin
/Library/Frameworks/Python.framework/Versions/3.11/bin
/opt/anaconda3/bin
/opt/homebrew/bin
/opt/homebrew/sbin
/System/Cryptexes/App/usr/bin
/Users/atatekeli/.cargo/bin
/Users/atatekeli/.config/yarn/global/node_modules/.bin
/Users/atatekeli/.dotnet/tools
/Users/atatekeli/.go/bin
/Users/atatekeli/.go/bin
/Users/atatekeli/.go/bin
/Users/atatekeli/.go/bin
/Users/atatekeli/.local/share/solana/install/active_release/bin
/Users/atatekeli/.yarn/bin
/Users/atatekeli/bin
/Users/atatekeli/Downloads/anaconda3/bin
/Users/atatekeli/opt/anaconda3/condabin
/bin
/sbin
/usr/bin
/usr/local/bin
/usr/local/go/bin
/usr/local/share/dotnet
/usr/sbin
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin

**Reply by Horpheyhemmy-18** [reply]
Run 
```
   /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
   
```
Once brew is installed 
Then run 

   ```
   brew install foundry

   ```

Foundry should be reinstalled

**Reply by Killpit** [reply]
I have brew installed, I consider deleting foundry packages forge, anvil, chisel, cast beforehand but I don't know if it's necessary to do so

**Reply by Horpheyhemmy-18** [reply]
Could be an option anyways.
But  ``` foundryup``` 
Should reinstall/update the package.

If it doesn't work,then  you could consider deleting the whole foundry package and reinstalling it

**Comment by Killpit** [comment]
Packages are down there

/Users/atatekeli/.cargo/bin, my plan is to remove them one by one, delete the project and reinstall foundry all over again, but I have to make sure none remains first

**Comment by Killpit** [comment]
Here are more details for this problem

```
macbook-pro:~ atatekeli$ cargo install --list
anvil v0.2.0 (https://github.com/foundry-rs/foundry#a5171a4a):
    anvil
avm v0.26.0 (https://github.com/coral-xyz/anchor#06c42327):
    anchor
    avm
chisel v0.2.0 (https://github.com/foundry-rs/foundry#a5171a4a):
    chisel
foundry-cli v0.2.0 (/Users/atatekeli/foundry/cli):
    cast
    forge
rustlings v5.2.1 (/Users/atatekeli/rustlings):
    rustlings
seahorse-lang v0.2.7:
    seahorse
trdelnik-cli v0.4.0:
    trdelnik

```

**Comment by drained69** [comment]
Specifying a particular version in your toml might help solve this. Example:
```
solc = "0.8.17"
```

**Reply by Killpit** [reply]
It didn't work

**Reply by drained69** [reply]
the repo you added to the question cannot be found.

**Reply by Killpit** [reply]
I deleted according to the Cyfrin team's advice and reinstalled foundry

**Comment by Killpit** [comment]
So what about using cargo command

**Comment by Killpit** [comment]
With cargo command according to documentation, but no foundryup 

```
forge 0.2.0 (71d8ea5 2024-01-09T14:39:42.577446000Z)
macbook-pro:~ atatekeli$ cast --version
cast 0.2.0 (71d8ea5 2024-01-09T14:42:29.595610000Z)
macbook-pro:~ atatekeli$ chisel --version
chisel 0.2.0 (71d8ea5 2024-01-09T14:45:10.956457000Z)
macbook-pro:~ atatekeli$ anvil --version
anvil 0.2.0 (71d8ea5 2024-01-09T14:47:33.371278000Z)
```

**Comment by Killpit** [comment]
@DevSwayam @Chinwuba22 @Horphey @tinchoabbate 

```
macbook-pro:~ atatekeli$ curl -L https://foundry.paradigm.xyz | bash
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  1942  100  1942    0     0   4032      0 --:--:-- --:--:-- --:--:--  4032
Installing foundryup...
######################################################################## 100.0%

Detected your preferred shell is bash and added foundryup to PATH. Run 'source /Users/atatekeli/.bashrc' or start a new terminal session to use foundryup.
Then, simply run 'foundryup' to install Foundry.
macbook-pro:~ atatekeli$ foundryup
-bash: foundryup: command not found

```

**Comment by Killpit** [comment]
![Ekran Resmi 2024-01-10 10 38 39](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/57031187/0ecc7579-b82d-46e0-aa5c-d1ea19280cfb)
![Ekran Resmi 2024-01-10 10 38 19](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/57031187/9813fa91-338d-4bc8-b566-af9fe00bd4c5)

New development, can you guys have a look

**Comment by drained69** [comment]
I don't think reinstalling foundry was the issue btw. Try changing the pragma version on all your contracts to >= 0.6.0 < 0.9.0.

**Comment by Killpit** [comment]
Right now, I uninstalled `cargo` packages for foundry and deleted `$PATH`s. I think I can start all over again. Will retry during installation.

**Comment by Killpit** [comment]
After reinstalling Foundry on macOS using foundryup, I can't run foundryup in the VS Code terminal. The error message says forge not found. I've reloaded VS Code and checked my PATH, but it's still not working. I'm using VS Code v1.55.0 and Foundry v0.2.0." with all the dependencies as the latest version, I just used the documentation, curl command, source and foundryup

![Ekran Resmi 2024-01-13 13 49 27](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/57031187/36f18202-74ee-48ca-acf3-5911a0c5829c)
![Ekran Resmi 2024-01-13 13 55 36](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/57031187/9271f1ce-9ca1-48a6-bd01-4b8f69fb2394)

**Reply by Killpit** [reply]
Just trying to make sure everything's alright, will look for further details later at the course, hope everything's alright

**Reply by cromewar** [reply]
were you able to solve this @Killpit ?

**Comment by nzlatev7** [comment]
I had a similar problem and was related to the imports, I checked your imports and can not see which dependecy cause this error!
